csrverifier/executable/csrverifier.go

```diff
@@ -0,0 +1,57 @@
+package executablecsrverifier
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+)
+
+const (
+	UserExecute os.FileMode = 1 << (6 - 3*iota)
+	GroupExecute
+	OtherExecute
+)
+
+func New(path string) (*ExecutableCSRVerifier, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, err
+	}
+
+	fileMode := fileInfo.Mode()
+	if fileMode.IsDir() {
+		return nil, errors.New("CSR Verifier executable is a directory")
+	}
+
+	filePerm := fileMode.Perm()
+	if filePerm&(UserExecute|GroupExecute|OtherExecute) == 0 {
+		return nil, errors.New("CSR Verifier executable is not executable")
+	}
+
+	return &ExecutableCSRVerifier{executable: path}, nil
+}
+
+type ExecutableCSRVerifier struct {
+	executable string
+}
+
+func (v *ExecutableCSRVerifier) Verify(data []byte) (bool, error) {
+	cmd := exec.Command(v.executable)
+
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return false, err
+	}
+
+	go func() {
+		defer stdin.Close()
+		stdin.Write(data)
+	}()
+
+	err = cmd.Run()
+	if err != nil {
+		// mask the executable error
```

Comment:
It will be hard to debug the reason this is failing if you hide the error. 
Add a logger to the struct and log the error here.

Reply:
The thing is "exit 1" is the standard behavior, if we want to stay compatible with the puppet interface. We could reroute the stdout / stderr of the script maybe.

Reply:
exit 1 is fine, but `cmd.Run` can fail for other reason. Imagine that I didn't set the executable flag on the binary or something. Masking errors should at least be logged because trying to understand why the run is failing for the user is tough, especially over github/slack.

Reply:
OK, will try to log it correctly.

BTW, The checks for the executable flags are done in `csrexecutableverifier.New`. I could still fail though.